RFCs/FS-1027-fsi-references.md

```diff
@@ -0,0 +1,100 @@
+# F# RFC FS-1027 - FSI Reference Model and extending `#r`
+
+* [x] Approved in principle
+* [ ] [FSLang Suggestion](https://github.com/fsharp/fslang-suggestions/issues/542)
+* [ ] Details: [under discussion](https://github.com/fsharp/fslang-design/issues/167)
+* [ ] Implementation: [In progress](https://github.com/Microsoft/visualfsharp/pull/FILL-ME-IN)
+
+# Summary
+[summary]: #summary
+
+The following extensions to `#r` are proposed:
+
+* `#r "impl: <path-to-implementation-assembly>"`
+* `#r "paket: <paket command>"`
+* `#r "nuget: <package-name>, <package-version>"`
+
+This extends the "language" of `#r` to support implementation and reference assemblies, NuGet packages, Paket dependencies, and potentially more.
+
+# Motivation
+[motivation]: #motivation
+
+Motivation for this change is twofold:
+
+1. There is a [strong desire](https://github.com/fsharp/fslang-suggestions/issues/542) to reference packages via `#r` instead of assemblies.  There has also been [experimental work](https://github.com/Microsoft/visualfsharp/pull/2483) to support Paket in this way.
+2. .NET Core uses both Reference Assemblies and Implementation Assemblies, which FSI must understand.  It does not understand this split today.
```

Comment:
@tmat could offer up more info, as I'm unaware of the plan for CSI.  It's a shared concern.  We're working with them here to make sure whatever we do makes sense.